src/cursor/find_cursor.ts

```diff
@@ -98,37 +98,50 @@ export class FindCursor<TSchema = any> extends ExplainableCursor<TSchema> {
   }
 
   /** @internal */
-  override async getMore(batchSize: number): Promise<CursorResponse> {
+  override async getMore(): Promise<CursorResponse> {
     const numReturned = this.numReturned;
-    if (numReturned) {
-      // TODO(DRIVERS-1448): Remove logic to enforce `limit` in the driver
-      const limit = this.findOptions.limit;
-      batchSize =
-        limit && limit > 0 && numReturned + batchSize > limit ? limit - numReturned : batchSize;
-
-      if (batchSize <= 0) {
-        try {
-          await this.close();
-        } catch (error) {
-          squashError(error);
-          // this is an optimization for the special case of a limit for a find command to avoid an
-          // extra getMore when the limit has been reached and the limit is a multiple of the batchSize.
-          // This is a consequence of the new query engine in 5.0 having no knowledge of the limit as it
-          // produces results for the find command.  Once a batch is filled up, it is returned and only
-          // on the subsequent getMore will the query framework consider the limit, determine the cursor
-          // is exhausted and return a cursorId of zero.
-          // instead, if we determine there are no more documents to request from the server, we preemptively
-          // close the cursor
-        }
-        return CursorResponse.emptyGetMore;
+    const limit = this.findOptions.limit ?? Infinity;
+    const remaining = limit - numReturned;
+
+    if (numReturned === limit && !this.id?.isZero()) {
+      // this is an optimization for the special case of a limit for a find command to avoid an
+      // extra getMore when the limit has been reached and the limit is a multiple of the batchSize.
+      // This is a consequence of the new query engine in 5.0 having no knowledge of the limit as it
+      // produces results for the find command.  Once a batch is filled up, it is returned and only
+      // on the subsequent getMore will the query framework consider the limit, determine the cursor
+      // is exhausted and return a cursorId of zero.
+      // instead, if we determine there are no more documents to request from the server, we preemptively
+      // close the cursor
+      try {
+        await this.close();
+      } catch (error) {
+        squashError(error);
       }
+      return CursorResponse.emptyGetMore;
+    }
+
+    // TODO(DRIVERS-1448): Remove logic to enforce `limit` in the driver
+    let cleanup: () => void = noop;
+    const { batchSize } = this.cursorOptions;
+    if (batchSize != null && batchSize > remaining) {
```

Comment:
I'm curious about the reasoning for this. What is the harm of sending a batch size greater than the number of remaining documents? Can this block just be removed?